source/Engine.cpp

```diff
@@ -1106,6 +1109,31 @@ void Engine::Click(const Point &from, const Point &to, bool hasShift)
 	this->hasShift = hasShift;
 	isRightClick = false;
 
+	// See if this click was in the ammoBox, then see if it clicked one of the icons.
+	if(ammoBox.Contains(from))
+	{
+		bool control = (SDL_GetModState() & KMOD_CTRL);
+		double ammoIconClickCenterX = (ammoBox.Right() + ammoBox.Left()) / 2.;
+		double posY = ammoBox.Bottom() - ammoPad;
+		for(const auto &it : ammo)
+		{
+			posY -= ICON_SIZE;
+			if(posY < ammoBox.Top() + ammoPad)
+				break;
+
+			double ammoIconClickCenterY = posY + ICON_SIZE / 2.;
+			Point ammoIconClickCenter(ammoIconClickCenterX, ammoIconClickCenterY);
+			Point ammoIconClickDimensions(AMMO_WIDTH, ICON_SIZE);
+			if(Rectangle(ammoIconClickCenter, ammoIconClickDimensions).Contains(from))
+			{
+				if(!control)
+					player.DeselectAllSecondaries();
+				player.ToggleAnySecondary(it.first);
+				return;
+			}
+		}
```

Comment:
Use a list of `ClickZone`s for this.

Reply:
That's what I was doing before. @petervdmeer suggested this instead. Should I change it back or something else entirely?

Reply:
Ah, I hadn't looked at the prior resolved commentary.

I think we're going to hit a design "divot" here, where MainPanel (owner of the engine instance) should support panels on top of itself that do not block the consideration of "is the player in flight?" along with panels that do (such as conversations, boarding, dialog, and hails). Flexible rendering of (perhaps interactive) model state information could then be removed from Engine and delegated appropriately via the UI event handling hierarchy. Consider the radar display, escorts, and ammo.

In this PR, we have the engine (which is decidedly _not_ a panel) needing to register UI elements to handle clicks and dispatch them to the player model. A more appropriate design would seem to be that `MainPanel`, upon construction / activation, adds a new "SecondaryWeaponAmmoDisplay" (name tbd) that is concerned only with registering the ClickZones and dispatching the changes to the player model when used.

I think the immediately actionable step here is to do something similar to what was done for the escort display: extract the logic of the ammo box into a class. That class can deduplicate the logic for knowing which icons are displayed, where exactly their click zones are, and dispatching the desired event to the player model.

Reply:
@warp-core Sorry for not reviewing in sufficient detail and sending you down a wrong path. tehhowch has a good point here.